### ISSUE TITLE
Design: Ego high availability

### ISSUE THREAD

**Issue by akachru-github** [icgc-argo/roadmap#520]
Currently, the RDPC is processing internally groomed/curated molecular data. We have full control over when and how this data is processed and can plan/react to any disruptions to the workflow system in stride. 

In the fall, ARGO will allow for external submissions of molecular data to the RDPC. As a part of this release, the workflow system should be a resilient as possible.

Various stages of our workflows depend on being able to authenticate with Ego. This ticket is meant to capture the design of either being highly available, perhaps by having the ability to gracefully failover to another instance hosted on Collab. 

The Requirement: Workflows being run in the RDPC should not be impacted by Platform availability. Ego is our only current dependency. 

Output of this ticket is a design doc. 

---
Documentation: 
https://wiki.oicr.on.ca/display/icgcargotech/Ego+HA

**Comment by akachru-github** [comment]
@christinayung / @rosibaj Question for you both:
 
I recall the spirit of this feature came from the OICR scheduled maintenance periods where ARGO Platform would be unavailable (including Ego). This ticket talks about Ego specifically. Are there other dependencies on the platform that we would need to make HA in order to make our workflow system completely un-affected by a scheduled maintenance period?

**Comment by akachru-github** [comment]
Suggestion: 
- Can this failover can be read-only as an MVP? We think it can be.

**Comment by rosibaj** [comment]
@akachru-github Yes, i think it can be read-only.  We dont need to add new uses, just have the users that are there able to login/have access tokens?

**Comment by akachru-github** [comment]
Thanks @rosibaj . 
A further note: @andricDu mentioned that once the Workflow Graph work is complete, there will be a Clinical dependency. We will need to keep this in mind if we want to maintain the RDPC integrity during a Platform outage (whether scheduled maintenance or disaster recovery scenario).

**Comment by akachru-github** [comment]
- Assuming second instance of Ego is READ ONLY

### Proposal: Secondary Ego
Can we have a secondary Ego run side by side with the primary Ego ?

- Primary Ego - Read / Write
- Secondary Ego - Read only
- Primary streams to Secondary 
- Secondary would be the instance used exclusively by RDPC
- dependencies: JWT generation, API key introspection
- Audit logging needs to be centralized. Need to ensure failover state still recording and centralizing logs from secondary instance. 
- Out of scope of this ticket but writing here so I don't forget: Ego logging should be revisited to ensure completeness. 

### Proposal: Cloud Hosted
- Cloud host Ego with load balancer
- AWS has a product that does this. 
- Costs the monies. 
- Datacenter should be in Toronto
- Platform / RDPC would point to the same LB. Either it gets routed to ARGO / RDPC. 
- Latency concerns for UI, but not for RDPC. 

Next steps:
- explore streamed secondary ego 
- discuss requirements beyond the RDPC for Ego HA.
